pkg/systemd/services/services.jsx

```diff
@@ -60,22 +61,42 @@ export const systemd_client = {
     user: cockpit.dbus(s_bus.BUS_NAME, { bus: "session" }),
 };
 const timedate_client = cockpit.dbus('org.freedesktop.timedate1');
-export let clock_realtime_now;
-export let clock_monotonic_now;
+export let clock_realtime_now = 0; // ms since epoch; call updateTime() before using
+let monotonic_timer_base = null; // µs
 
 export const MAX_UINT64 = 2 ** 64 - 1;
 
 export function updateTime() {
-    cockpit.spawn(["cat", "/proc/uptime"])
-            .then(function(contents) {
-                // first number is time since boot in seconds with two fractional digits
-                const uptime = parseFloat(contents.split(' ')[0]);
-                clock_monotonic_now = parseInt(uptime * 1000000, 10);
-            }, ex => console.log(ex.toString()));
-    cockpit.spawn(["date", "+%s"])
-            .then(time => {
-                clock_realtime_now = parseInt(time, 10) * 1000;
-            }, ex => console.log(ex.toString()));
+    /* To correctly interpret monotonic timers, we need to read CLOCK_MONOTONIC. This cannot be read with shell tools,
+     * and the only source in /proc is /proc/timer_list which is readable by root only (and also rather hard to parse). */
+    python.spawn("import time; print(time.clock_gettime_ns(time.CLOCK_REALTIME), time.clock_gettime_ns(time.CLOCK_MONOTONIC))",
+                 null, { err: "message" })
+            .then(output => {
+                const [realtime, monotonic] = output.split(' ').map(ns => parseInt(ns));
+                clock_realtime_now = realtime / 1000000;
+                monotonic_timer_base = (realtime - monotonic) / 1000;
+            })
+            .catch(ex => {
+                /* If Python is not available, fall back to reading CLOCK_BOOTTIME from /proc/uptime. This will
+                 * give incorrect results if the machine was suspended during the current boot, but it's the
+                 * best approximation that we can get. */
+                console.log("Failed to read clocks with Python, using boot time fallback:", ex.toString());
+                Promise.all([
+                    cockpit.file("/proc/uptime").read(),
+                    cockpit.spawn(["date", "+%s"])
+                ])
+                        .then(([uptime_contents, date_output]) => {
+                            clock_realtime_now = parseInt(date_output, 10) * 1000;
```

Comment:
These 11 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18466-20230308-091051-1f44fdef-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

Reply:
They actually would be on our fcos image, as that doesn't have Python.

Reply:
What?  That's insanity.

Reply:
@mvollmer : You mean CoreOS not having Python? Consciuous decision on their side as they don't like Ansible. Or in case you didn't mean that: I wasn't suggesting to actually run/expose coverage on fcos (we skip half of the tests there, after all) -- it was just an explanation.

Reply:
> @mvollmer : You mean CoreOS not having Python? Consciuous decision on their side as they don't like Ansible.

That's what I meant. Hmm.